src/tls/test-cockpit-certificate-ensure.c

```diff
@@ -0,0 +1,432 @@
+  g_assert_no_error (error);
+
+  self->runtime_dir_fd = open (self->runtime_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+  g_assert_no_errno (self->runtime_dir_fd);
+
+  /* config dir */
+  self->config_dir = g_dir_make_tmp ("cockpit-test-config.XXXXXX", &error);
+  g_assert_no_error (error);
+
+  self->config_dir_fd = open (self->config_dir, O_PATH | O_DIRECTORY | O_NOFOLLOW | O_CLOEXEC);
+  g_assert_no_errno (self->config_dir_fd);
+
+  int r = mkdirat (self->config_dir_fd, "cockpit", 0700);
+  g_assert_no_errno (r);
+
+  r = mkdirat (self->config_dir_fd, "cockpit/ws-certs.d", 0700);
+  g_assert_no_errno (r);
+
+  /* populate ws-certs.d */
+  for (int i = 0; tc->files[i]; i++)
+    {
+      g_autofree gchar *linkname = g_strconcat ("cockpit/ws-certs.d", strrchr (tc->files[i], '/'), NULL);
+      r = symlinkat (tc->files[i], self->config_dir_fd, linkname);
+      g_assert_no_errno (r);
+    }
+
+  /* launcher */
+  self->launcher = g_subprocess_launcher_new (G_SUBPROCESS_FLAGS_STDOUT_PIPE | G_SUBPROCESS_FLAGS_STDERR_PIPE);
+  g_subprocess_launcher_setenv (self->launcher, "XDG_CONFIG_DIRS", self->config_dir, TRUE);
+  g_subprocess_launcher_setenv (self->launcher, "RUNTIME_DIRECTORY", self->runtime_dir, TRUE);
+}
+
+static void
+test_check (Fixture *fixture,
+            gconstpointer data)
+{
+  const TestCase *tc = data;
+  g_autoptr(GError) error = NULL;
+
+  g_autoptr(GSubprocess) helper = g_subprocess_launcher_spawn (fixture->launcher, &error,
+                                                               CERTIFICATE_HELPER, "--check", NULL);
+  g_assert_no_error (error);
+
+  g_autofree gchar *stdout_str = NULL;
+  g_autofree gchar *stderr_str = NULL;
+  gboolean result = g_subprocess_communicate_utf8 (helper, NULL, NULL, &stdout_str, &stderr_str, &error);
+  g_assert_no_error (error);
+  g_assert (result);
+
+  g_assert (g_subprocess_get_if_exited (helper));
+
+  cockpit_assert_strmatch (stdout_str, tc->check_stdout);
+  cockpit_assert_strmatch (stderr_str, tc->check_stderr);
+  g_assert_cmpint (g_subprocess_get_exit_status (helper), ==, tc->check_exit);
+}
+
+static void
+test_copy (Fixture *fixture,
+           gconstpointer data)
+{
+  const TestCase *tc = data;
+  g_autoptr(GError) error = NULL;
+
+  g_autoptr(GSubprocess) helper = g_subprocess_launcher_spawn (fixture->launcher, &error,
+                                                               CERTIFICATE_HELPER, "--for-cockpit-tls", NULL);
+  g_assert_no_error (error);
+
+  g_autofree gchar *stdout_str = NULL;
+  g_autofree gchar *stderr_str = NULL;
+  gboolean result = g_subprocess_communicate_utf8 (helper, NULL, NULL, &stdout_str, &stderr_str, &error);
+  g_assert_no_error (error);
+  g_assert (result);
+
+  g_assert (g_subprocess_get_if_exited (helper));
+
+  cockpit_assert_strmatch (stdout_str, tc->copy_stdout);
+  cockpit_assert_strmatch (stderr_str, tc->copy_stderr);
+  g_assert_cmpint (g_subprocess_get_exit_status (helper), ==, tc->copy_exit);
```

Comment:
This whole code is identical to test_check() aside from the `--for-cockpit-tls` and the duplication of/redundant copy/check stdout/err. I propose to move that extra option into TestCase, then their definitions would also make it clear which mode they apply to; and factorize this code with test_check().

(Same, this can be a follow-up)

Reply:
Ya.  I also got this feeling when I was writing it.  This could be controlled by a flag in the testcase struct.